File: src/main/java/beam/agentsim/events/handling/BeamEventsWriterXML.java
package beam.agentsim.events.handling;

import beam.agentsim.events.LoggerLevels;
import beam.sim.BeamServices;
import org.matsim.api.core.v01.events.Event;
import org.matsim.core.utils.io.UncheckedIOException;

import java.io.IOException;
import java.util.Map;

/**
 * BEAM
 */
public class BeamEventsWriterXML extends BeamEventsWriterBase{
    public BeamEventsWriterXML(String outfilename, BeamEventsLogger beamEventLogger, BeamServices beamServices, Class<?> eventTypeToLog) {
        super(outfilename, beamEventLogger, beamServices, eventTypeToLog);
        writeHeaders();
    }

    @Override
    public void writeHeaders(){
        try {
            this.out.write("<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<events version=\"1.0\">\n");
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public void closeFile() {
        try {
            this.out.write("</events>");
            // I added a "\n" to make it look nicer on the console.  Can't say if this may have unintended side
            // effects anywhere else.  kai, oct'12
            // fails signalsystems test (and presumably other tests in contrib/playground) since they compare
            // checksums of event files.  Removed that change again.  kai, oct'12
            this.out.close();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Deprecated
    public void init(final String outfilename) {
        throw new RuntimeException("Please create a new instance.");
    }

    @Override
    public void reset(final int iter) {
    }

    @Override
    protected void writeEvent(final Event event) {
        if(! (beamEventLogger.getLoggingLevel(event) == LoggerLevels.OFF)){
            try{
                this.out.append("\t<event ");
                Map<String, String> attr = beamEventLogger.getAttributes(event);
                for (Map.Entry<String, String> entry : attr.entrySet()) {
                    this.out.append(entry.getKey());
                    this.out.append("=\"");
                    this.out.append(encodeAttributeValue(entry.getValue()));
                    this.out.append("\" ");
                }

             this.out.append(" />\n");
//			this.out.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    // the following method was taken from MatsimXmlWriter in order to correctly encode attributes, but
    // to forego the overhead of using the full MatsimXmlWriter.
    /**
     * Encodes the given string in such a way that it no longer contains
     * characters that have a special meaning in xml.
     *
     * @see <a href="http://www.w3.org/International/questions/qa-escapes#use">http://www.w3.org/International/questions/qa-escapes#use</a>
     * @param attributeValue
     * @return String with some characters replaced by their xml-encoding.
     */
    private String encodeAttributeValue(final String attributeValue) {
        if (attributeValue == null) {
            return null;
        }
        int len = attributeValue.length();
        boolean encode = false;
        for (int pos = 0; pos < len; pos++) {
            char ch = attributeValue.charAt(pos);
            if (ch == '<') {
                encode = true;
                break;
            } else if (ch == '>') {
                encode = true;
                break;
            } else if (ch == '\"') {
                encode = true;
                break;
            } else if (ch == '&') {
                encode = true;
                break;
            }
        }
        if (encode) {
            StringBuffer bf = new StringBuffer();
            for (int pos = 0; pos < len; pos++) {
                char ch = attributeValue.charAt(pos);
                if (ch == '<') {
                    bf.append("&lt;");
                } else if (ch == '>') {
                    bf.append("&gt;");
                } else if (ch == '\"') {
                    bf.append("&quot;");
                } else if (ch == '&') {
                    bf.append("&amp;");
                } else {
                    bf.append(ch);
                }
            }

            return bf.toString();
        }
        return attributeValue;

    }
}

File: src/main/scala/beam/agentsim/events/handling/BeamEventsWriterXML.scala
package beam.agentsim.events.handling

import beam.sim.BeamServices
import org.matsim.api.core.v01.events.Event

/**
  * @author Bhavya Latha Bandaru.
  * Helper class to write BEAM events to an external file.
  * @param outFileName writer instance to write to external file
  * @param beamEventLogger beam events logger
  * @param beamServices beam services
  * @param eventTypeToLog type of event to log
  */
class BeamEventsWriterXML(
  var outFileName: String,
  beamEventLogger: BeamEventsLogger,
  beamServices: BeamServices,
  eventTypeToLog: Class[_]
) extends BeamEventsWriterBase(outFileName, beamEventLogger, beamServices, eventTypeToLog) {

  val specialCharToEscape: Map[Char, String] = Map('<' -> "&lt;", '>' -> "&gt;", '"' -> "&quot;", '&' -> "&amp;")
  val specialChars: Array[Char] = specialCharToEscape.keys.toArray

  val header: String =
    """<?xml version="1.0" encoding="utf-8"?>
<events version="1.0">""".stripMargin

  writeHeaders()

  /**
    * Writes the events to the xml file.
    *
    * @param event event to written
    */
  override protected def writeEvent(event: Event): Unit = {
    //get all the event attributes
    outWriter.append("\t<event ")
    event.getAttributes.forEach((name, value) => {
      outWriter.append(name)
      outWriter.append("=\"")
      outWriter.append(encodeAttributeValue(value))
      outWriter.append("\" ")
    })
    outWriter.append("/>")
    outWriter.newLine()
  }

  /**
    * Adds the xml header to the file.
    */
  override protected def writeHeaders(): Unit = {
    outWriter.write(header)
    outWriter.newLine()
  }

  /**
    * Appends the footer and closes the file stream.
    */
  override def closeFile(): Unit = {
    outWriter.write("</events>")
    outWriter.close()
  }

  /**
    * Encodes any special character encountered in the xml attribute value.
    *
    * @param attributeValue xml attribute value
    * @return encoded attribute value
    */
  private def encodeAttributeValue(attributeValue: String): String = {
    var i: Int = 0
    var result = attributeValue
    while (i < specialChars.length) {
      val char = specialChars(i)
      if (result.indexOf(char) >= 0)
        result = result.replace(char.toString, specialCharToEscape(char))
      i += 1
    }
    result
  }
}
